### DISCUSSION TITLE
MySQL 索引设计概要 · /sql-index-intro

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#249] [Announcements]
https://draveness.me/sql-index-intro

**Comment by CodingCrush** [comment]
学习了

**Comment by null2046** [comment]
SELECT id, name, age FROM users
WHERE age BETWEEN 18 AND 21
  AND city = "Beijing"
ORDER BY name;

这个age 条件 mysql会优化为 age in (18,19,20,21) 并不是一个范围查询了吧。

所以(city,age,name,id) 还是个三星索引吧？

**Comment by draveness** [comment]
> @zhengvictor
> SELECT id, name, age FROM users
> WHERE age BETWEEN 18 AND 21
>   AND city = "Beijing"
> ORDER BY name;
> 
> 这个age 条件 mysql会优化为 age in (18,19,20,21) 并不是一个范围查询了吧。
> 
> 所以(city,age,name,id) 还是个三星索引吧？

我也不知道，你 explain 一下看看是不是用了 index 吧

**Comment by Hugo-Gao** [comment]
写的真好，向博主学习

**Comment by jacentsao** [comment]
请教一下绘图使用的是什么软件，谢谢！

**Comment by leezhijie** [comment]
同问使用什么软件做的图

**Comment by coderxiaoyang** [comment]
图片使用 Sketch 进行绘制 作者每篇文章尾部都有写

**Comment by draveness** [comment]
> @lufeisan
> 图片使用 Sketch 进行绘制 作者每篇文章尾部都有写

是啊，我真的是不想回复...

**Comment by Xunzi229** [comment]
假设字段 在组合索引中,  然后又增加该字段单独索引, 这样会有用处吗？

**Comment by draveness** [comment]
> 假设字段 在组合索引中, 然后又增加该字段单独索引, 这样会有用处吗？

没明白你想问啥

**Comment by Xunzi229** [comment]
> @draveness
> > 假设字段 在组合索引中, 然后又增加该字段单独索引, 这样会有用处吗？
> 
> 没明白你想问啥

假设在 一个表中 增加索引 
add_index :users, :name
add_index :users, [:name, :age]
那么这样在 单独查询 name 的时候， 使用的是 组合索引 还是 单独的那个 name 的索引 ？

**Comment by draveness** [comment]
> > @draveness
> > > 假设字段 在组合索引中, 然后又增加该字段单独索引, 这样会有用处吗？
> > 
> > 
> > 没明白你想问啥
> 
> 假设在 一个表中 增加索引
> add_index :users, :name
> add_index :users, [:name, :age]
> 那么这样在 单独查询 name 的时候， 使用的是 组合索引 还是 单独的那个 name 的索引 ？

不知道会用哪个，不过第二个索引会『覆盖』第一索引，第一个索引是冗余的，没什么用

**Comment by fxyJAVA** [comment]
这里的窄索引和宽索引就是指的是非聚族与聚族索引吧，还是第一次听见窄宽这两个概念

**Comment by draveness** [comment]
宽和窄说的是索引中数量的多少

**Comment by fxyJAVA** [comment]
> @draveness
> 宽和窄说的是索引中数量的多少
> On Aug 22, 2019, 11:46 PM +0800, black crow <notifications@github.com>, wrote:
> > 这里的窄索引和宽索引就是指的是非聚族与聚族索引吧，还是第一次听见窄宽这两个概念
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub, or mute the thread.
> 
孤陋寡闻，学习了。昨晚没看太仔细，先入为主了

**Comment by gaotuan** [comment]
"如果在数据库的缓存池中没有找到对应的数据页，那么会去内存中寻找对应的页面" 
这一句不太理解，如果要查询的page不在bufferPoll里面，那么会访问磁盘，然后把数据缓存到bufferPoll。
如果不在缓存池，会去内存中找对应的页面，这个内存指的是哪里？

**Comment by draveness** [comment]
> "如果在数据库的缓存池中没有找到对应的数据页，那么会去内存中寻找对应的页面"
> 这一句不太理解，如果要查询的page不在bufferPoll里面，那么会访问磁盘，然后把数据缓存到bufferPoll。
> 如果不在缓存池，会去内存中找对应的页面，这个内存指的是哪里？

不太理解你的问题，你可以理解内存不就是内存么...

**Reply by TuDouZi** [reply]
因为你没有说清楚，你应该这么解释：buffer pool里会缓存索引页和存储数据记录的页，当只有索引页没有数据记录的页的时候需要一次磁盘io读到内存，当索引页也没有的时候，会随机io很多次读磁盘

**Reply by TuDouZi** [reply]
我理解错了，又重新看了一下， 这里的内存应该是操作系统的物理内存，缓冲池只是mysql应用程序的进程内存空间

**Comment by hhyvs111** [comment]
> @Xunzi229
> > @draveness
> > > 假设字段 在组合索引中, 然后又增加该字段单独索引, 这样会有用处吗？
> > 
> > 没明白你想问啥
> 
> 假设在 一个表中 增加索引 
> add_index :users, :name
> add_index :users, [:name, :age]
> 那么这样在 单独查询 name 的时候， 使用的是 组合索引 还是 单独的那个 name 的索引 ？ 

在建第二个的时候，它就会生成第一个索引

**Comment by maxnoodles** [comment]
从公众号过来的，文章写得非常好！
不好意思，看完文章有 3 个问题打扰您:
1.  这套关系数据库 索引方案在 MONGODB 这样非关系型数据库的单集合适用吗？
2.  以前学校的书还有很多文章都说，频繁改动的列不适用索引，但是查询有需要对这一列进行排序，该如何取舍(例如:更新时间)
3.  如果为了获取 三星索引 的第三颗 ，对过多的字段建立索引带来的收益会不会很低，博主平时建立索引所需最低收益大概是多少。

**Comment by draveness** [comment]
> 1. 这套关系数据库 索引方案在 MONGODB 这样非关系型数据库的单集合适用吗？

只能说部分使用，MongoDB 在数据之间的关系设计上与 MySQL 非常不同

> 2. 以前学校的书还有很多文章都说，频繁改动的列不适用索引，但是查询有需要对这一列进行排序，该如何取舍(例如:更新时间)

直接用就行了，到了复杂场景也不会问出这个问题。

> 3. 如果为了获取 三星索引 的第三颗 ，对过多的字段建立索引带来的收益会不会很低，博主平时建立索引所需最低收益大概是多少。

一般情况下，两星就够了，只要我们能限制一次查询出来的数据量。

**Comment by maxnoodles** [comment]
> @draveness
> > 1. 这套关系数据库 索引方案在 MONGODB 这样非关系型数据库的单集合适用吗？
> 
> 只能说部分使用，MongoDB 在数据之间的关系设计上与 MySQL 非常不同
> 
> > 2. 以前学校的书还有很多文章都说，频繁改动的列不适用索引，但是查询有需要对这一列进行排序，该如何取舍(例如:更新时间)
> 
> 直接用就行了，到了复杂场景也不会问出这个问题。
> 
> > 3. 如果为了获取 三星索引 的第三颗 ，对过多的字段建立索引带来的收益会不会很低，博主平时建立索引所需最低收益大概是多少。
> 
> 一般情况下，两星就够了，只要我们能限制一次查询出来的数据量。

学习了，非常感谢您的解答。

**Comment by bush2582** [comment]
我同样在看这本书，书里面的方法论不错。但是一些方法的所适应的硬件场景已经有点过时。不知道博主对这块有没有做什么改进或者深入的工作？

**Comment by bush2582** [comment]
这本书有些地方看的我有点迷糊，如果有可能，希望能和博主交流下~~~

**Comment by ichengzi** [comment]
老实说搞定索引其实很不容易，文中例子其实简化了很多。 

当业务太复杂时， 挑选合适的索引可能还要业务的适当妥协， 尤其是那种后台查询表单，一堆查询条件，组合起来很复杂。 

其实不管几星索引， 让索引起到索引本来的作用，尽量缩小查询返回，减少回表次数，差不多就能满足性能要求了。

> 归根到底， 数据库的核心矛盾就是磁盘随机读写能力有限， 硬件限制没有办法，只能尽量减少随机读写。

**Comment by gczgcz2015** [comment]
> 第一颗星需要取出所有等值谓词中的列，作为索引开头的最开始的列（任意顺序）；

这里的“任意顺序”有点问题吧，应该把过滤因子越小的放到越前面这样可以尽早的过滤掉数据。

**Comment by zyj123** [comment]
缓冲池和内存有什么区别，缓冲池不就是在内存里吗？

**Reply by draveness** [reply]
+ 缓存池更侧重说一种重复利用对象或者预先分配的机制
+ 内存更侧重的是一种存储介质，与磁盘、寄存器这些相对应
